cve_bin_tool/checkers/__init__.py

```diff
@@ -284,10 +285,10 @@ def __new__(cls, name, bases, props):
 
 
 class Checker(metaclass=CheckerMetaClass):
-    CONTAINS_PATTERNS: list = []
-    VERSION_PATTERNS: list = []
-    FILENAME_PATTERNS: list = []
-    VENDOR_PRODUCT: list = []
+    CONTAINS_PATTERNS: list[str] = []
+    VERSION_PATTERNS: list[str] = []
+    FILENAME_PATTERNS: list[str] = []
+    VENDOR_PRODUCT: Optional[list[tuple[str, str]]] = []
```

Comment:
Why did you type it as `Optional`? In what case can it be `None`? From glancing at the code it seems to me that this list has to include at least 1 pair in checker implementations, no checkers set it to `None`.

Reply:
I was basing it off https://github.com/intel/cve-bin-tool/blob/205f43822c74bae3b0863ac11ed7a7426d56d62f/cve_bin_tool/helper_script.py#L56.  I'm not yet up to speed on the checker implementations (was going to tackle implementing a new checker next) so if you think it's not correct I can just set it to a list of tuples and drop `None`.  If that's the case, should typing on L56 in helper_script.py also be updated?

Reply:
...and actually maybe `find_vendor_product()` would need to be updated to return empty lists instead of `None`?

Reply:
I suspect that checkers with `None` instead of the list will just error out, check these lines:
[1](https://github.com/intel/cve-bin-tool/blob/main/cve_bin_tool/version_scanner.py#L235)
```py
for vendor, product in checker.VENDOR_PRODUCT:
```
[2](https://github.com/intel/cve-bin-tool/blob/205f43822c74bae3b0863ac11ed7a7426d56d62f/cve_bin_tool/checkers/__init__.py#L263-L265)
```py
# Validate that we have at least one vendor product pair
if len(cls.VENDOR_PRODUCT) < 1:
    raise AssertionError(f"{name} needs at least one vendor product pair")
```

Both will be runtime errors if it's `None`.

Reply:
Agreed, that should never be `None` but empty list is okay?  Any reason not to `return []` in those lines in `find_vendor_product()`?

Also for the `re.Pattern` type issue I agree it isn't clear here how to type this but certainly from a developer checker implementation they submit them as strings right so `list[str]` I think is still okay even though it does get transformed into an re.Pattern afterwards.   I want to review the code more a bit and see if I have other ideas but happy to keep pushing to get this done.